util/auto_skip_compressor.cc

```diff
@@ -123,9 +123,202 @@ std::unique_ptr<Compressor> AutoSkipCompressorManager::GetCompressorForSST(
+  // Create cost predictors for each compression type and level
+  wa->cost_predictors.reserve(compression_levels_.size());
+  for (size_t i = 0; i < compression_levels_.size(); i++) {
+    CompressionType type_ = static_cast<CompressionType>(i + 1);
+    if (compression_levels_[type_ - 1].size() == 0) {
+      wa->cost_predictors.emplace_back();
+      continue;
+    } else {
+      std::vector<IOCPUCostPredictor*> predictors_diff_levels;
+      predictors_diff_levels.reserve(compression_levels_[type_ - 1].size());
+      for (size_t j = 0; j < compression_levels_[type_ - 1].size(); j++) {
+        predictors_diff_levels.emplace_back(new IOCPUCostPredictor(10));
+      }
+      wa->cost_predictors.emplace_back(std::move(predictors_diff_levels));
+    }
+  }
+  return ManagedWorkingArea(wa, this);
+}
+void CPUIOAwareCompressor::ReleaseWorkingArea(WorkingArea* wa) {
+  // remove all created cost predictors
+  for (auto& prdictors_diff_levels :
+       static_cast<CPUIOAwareWorkingArea*>(wa)->cost_predictors) {
+    for (auto& predictor : prdictors_diff_levels) {
+      delete predictor;
+    }
+  }
+  delete static_cast<CPUIOAwareWorkingArea*>(wa);
+}
+
+Status CPUIOAwareCompressor::CompressBlockAndRecord(
+    size_t choosen_compression_type, size_t compresion_level_ptr,
+    Slice uncompressed_data, std::string* compressed_output,
+    CompressionType* out_compression_type, CPUIOAwareWorkingArea* wa) {
+  assert(choosen_compression_type < allcompressors_.size());
+  assert(compresion_level_ptr <
+         allcompressors_[choosen_compression_type].size());
+  StopWatchNano timer(Env::Default()->GetSystemClock().get(), true);
+  Status status =
+      allcompressors_[choosen_compression_type][compresion_level_ptr]
+          ->CompressBlock(uncompressed_data, compressed_output,
+                          out_compression_type, &(wa->wrapped));
+  auto output_length = compressed_output->size();
+  TEST_SYNC_POINT_CALLBACK(
+      "CPUIOAwareCompressor::CompressBlockAndRecord::"
+      "DelaySetCompressedOutputSize",
+      &output_length);
+  auto predictor =
+      wa->cost_predictors[choosen_compression_type][compresion_level_ptr];
+  auto cpu_time = timer.ElapsedNanos();
+  TEST_SYNC_POINT_CALLBACK(
+      "CPUIOAwareCompressor::CompressBlockAndRecord::"
+      "GetCPUTime",
+      &cpu_time);
+  TEST_SYNC_POINT_CALLBACK(
+      "CPUIOAwareCompressor::CompressBlockAndRecord::"
+      "GetCompressedOutputSize",
+      &output_length);
+  predictor->CPUPredictor.Record(cpu_time);
+  predictor->IOPredictor.Record(output_length);
+  TEST_SYNC_POINT_CALLBACK(
+      "CPUIOAwareCompressor::CompressBlockAndRecord::GetIOPredictor",
+      &(wa->cost_predictors[choosen_compression_type][compresion_level_ptr]
+            ->IOPredictor));
+  TEST_SYNC_POINT_CALLBACK(
+      "CPUIOAwareCompressor::CompressBlockAndRecord::GetCPUPredictor",
+      &(wa->cost_predictors[choosen_compression_type][compresion_level_ptr]
+            ->CPUPredictor));
```

Comment:
Can we be more careful in using sync points such as combining them into one by passing a pair, tuple or list etc instead of having separate sync points, each for passing one piece of info?